tests/pems/crl/generate.sh

```diff
@@ -0,0 +1,281 @@
+pushd "${base_dir}" || exit
+
+#
+# Generate certificate chains of valid and revoked certificates with 1 intermediate CA.
+#
+# root -> valid intermediate   -> valid leaf
+#                              -> revoked leaf
+#      -> revoked intermediate -> valid leaf
+#                              -> revoked leaf
+#
+
+# Generate root certificate
+init_ca_dir root
+generate_key_cert root ca
+
+# Generate intermediate certificates
+init_ca_dir root/intermediate
+generate_key_cert root/intermediate intermediate
+sign_cert root root/intermediate ca
+
+init_ca_dir root/intermediate_revoked
+generate_key_cert root/intermediate_revoked intermediate
+sign_cert root root/intermediate_revoked ca
+
+# Generate leaf certificates
+mkdir root/intermediate/leaf
+generate_key_cert root/intermediate/leaf leaf
+sign_cert root/intermediate root/intermediate/leaf intermediate
+
+mkdir root/intermediate/leaf_revoked
+generate_key_cert root/intermediate/leaf_revoked leaf
+sign_cert root/intermediate root/intermediate/leaf_revoked intermediate
+
+mkdir root/intermediate_revoked/leaf
+generate_key_cert root/intermediate_revoked/leaf leaf
+sign_cert root/intermediate_revoked root/intermediate_revoked/leaf intermediate
+
+mkdir root/intermediate_revoked/leaf_revoked
+generate_key_cert root/intermediate_revoked/leaf_revoked leaf
+sign_cert root/intermediate_revoked root/intermediate_revoked/leaf_revoked intermediate
+
+# Generate CRLs and revoke the revoked certificates
+generate_crl root
+revoke_cert root root/intermediate_revoked
+
+generate_crl root/intermediate
+revoke_cert root/intermediate root/intermediate/leaf_revoked
+
+generate_crl root/intermediate_revoked
+revoke_cert root/intermediate_revoked root/intermediate/leaf_revoked
+
+# Ensure openssl verify reads generated CRLs and properly rejects certificates
+verify_cert root/chain.pem root/intermediate/cert.pem root/crl.pem 0
+verify_cert root/chain.pem root/intermediate_revoked/cert.pem root/crl.pem 1
+
+verify_cert root/intermediate/chain.pem root/intermediate/leaf/cert.pem root/intermediate/crl.pem 0
+verify_cert root/intermediate/chain.pem root/intermediate/leaf_revoked/cert.pem root/intermediate/crl.pem 1
+
+verify_cert root/intermediate_revoked/chain.pem root/intermediate_revoked/leaf/cert.pem root/intermediate_revoked/crl.pem 0
+verify_cert root/intermediate_revoked/chain.pem root/intermediate_revoked/leaf_revoked/cert.pem root/intermediate_revoked/crl.pem 1
+
+# Generate CRLs with invalid timestamps. Requires Openssl 3.0.
+ca_dir="root/intermediate" \
+openssl ca -config "${openssl_conf_path}" -gencrl \
+    -crl_lastupdate 21220101000000Z \
+    -out "${install_dir}/intermediate_invalid_last_update_crl.pem" \
+    || exit
+
+ca_dir="root/intermediate" \
+openssl ca -config "${openssl_conf_path}" -gencrl \
+    -crl_nextupdate 19220101000000Z \
+    -out "${install_dir}/intermediate_invalid_next_update_crl.pem" \
+    || exit
+
+#
+# Create certificate chain pems and copy CRLs and keys
```

Comment:
Yeah it's only in this last section that I feel like I really start to struggle with the names you've chosen for your pem files. What the heck is a valid_valid_key 😆

Reply:
Fair enough... I wanted some way to indicate which certificate in the chain is revoked, and which one is valid. Would unrevoked_unrevoked_cert_chain/unrevoked_revoked_cert_chain be better? Do you know of a better way to name them?

Reply:
Well, maybe explain to me what you mean by valid_valid_key. I'm guessing it's a key for a certificate in a chain where the intermediate and leaf certificates are not revoked? But even then, I don't know which certificate the key is for; the leaf or the intermediate?

Reply:
Oh got it. Yeah, it's the key for the leaf certificate in the valid -> valid chain. Would `valid_valid_leaf_key.pem` make more sense?

Reply:
I did generate a key in order to create the intermediate cert, but I didn't copy that up in the script or commit it, since I didn't think we would need it for anything. I wanted the leaf cert to give to s2nd to test sending the certificate chain to s2nc. Is there a reason to also copy the intermediate key up?

Reply:
No, but then I would tie the name of the key file with which certificate it was generated from. What about "valid_valid_leaf_key.pem", "valid_invalid_leaf_key.pem" ...etc. This is my best suggestion 🤷‍♀️

Reply:
Sounds good! Added: d92043ce9eb202b595ac1a1c6dd8b0a0ed08e611

Reply:
I'm still not a fan of the names. Can we be more specific? I think the increased length of the cert name will be worth it.

Like:
leaf_revoked_(cert_chain|key).pem
intermediate_revoked_(cert_chain|key).pem
all_revoked_(cert_chain|key).pem

We also really don't need to make sure that we specify that keys belong to the leaf certs-- that's just how certs work in TLS :)

Reply:
Updated: 2e906846dacc5bfbeb81a68be2f887fd1616ae67

Reply:
Names are so hard :(